lib/src/resolve_source.cc

```
@@ -1,14 +1,102 @@
+    }
+
+    resolve_source::result_with_path resolve_source::lookup_subst(resolve_context context,
+                                                                  std::shared_ptr<substitution_expression> subst,
+                                                                  int prefix_length) const {
+
+        auto result = find_in_object(_root, context, subst->get_path());
+
+        if (!result.result.value) {
+            auto unprefixed = subst->get_path().sub_path(prefix_length);
+
+            if (prefix_length > 0) {
+                result = find_in_object(_root, result.result.context, unprefixed);
+            }
+
+            if(!result.result.value && result.result.context.options().get_use_system_environment()) {
+                // TODO: implement this o:
+                throw config_exception("lookup in system environment is not implemented");
+            }
+        }
+
+        return result;
+
+    }
+
+    resolve_source::result_with_path resolve_source::find_in_object(shared_object obj, resolve_context context,
+                                                                    path the_path) const {
+        auto restriction = context.restrict_to_child();
+        auto partially_resolved = context.restrict(the_path).resolve(dynamic_pointer_cast<const config_value>(obj), {obj});
+        auto new_context = partially_resolved.context.restrict(restriction);
+
+        if(auto value = dynamic_pointer_cast<const config_object>(partially_resolved.value)) {
+            value_with_path pair = find_in_object(value, the_path);
+            return {make_resolve_result(new_context, pair.value), pair.path_from_root};
+
+        } else {
+            throw bug_or_broken_exception("resolved object to non-object");
+        }
+
+    }
+
+    resolve_source::value_with_path resolve_source::find_in_object(shared_object obj, path the_path) const {
+        // TODO: implement improve_not_resolved
+        return find_in_object(obj, the_path, {});
```

Should we throw an exception here?